File: lib/src/main/java/com/hootsuite/nachos/validator/ChipifyingNachoValidator.java
package com.hootsuite.nachos.validator;

import android.support.annotation.NonNull;
import android.text.SpannableStringBuilder;
import android.util.Pair;

import com.hootsuite.nachos.tokenizer.ChipTokenizer;

import java.util.List;

/**
 * A {@link NachoValidator} that deems text to be invalid if it contains
 * unterminated tokens and fixes the text by chipifying all the unterminated tokens.
 */
public class ChipifyingNachoValidator implements NachoValidator {

    @Override
    public boolean isValid(@NonNull ChipTokenizer chipTokenizer, CharSequence text) {

        // The text is considered valid if there are no unterminated tokens (everything is a chip)
        List<Pair<Integer, Integer>> unterminatedTokens = chipTokenizer.findAllTokens(text);
        return unterminatedTokens.isEmpty();
    }

    @Override
    public CharSequence fixText(@NonNull ChipTokenizer chipTokenizer, CharSequence invalidText) {
        SpannableStringBuilder newText = new SpannableStringBuilder(invalidText);
        chipTokenizer.terminateAllTokens(newText);
        return newText;
    }
}

File: nachos/src/main/java/com/hootsuite/nachos/validator/ChipifyingNachoValidator.kt
package com.hootsuite.nachos.validator

import android.text.SpannableStringBuilder
import androidx.annotation.NonNull
import com.hootsuite.nachos.tokenizer.ChipTokenizer

/**
 * A [NachoValidator] that marks text invalid when unterminated tokens are present and
 * fixes it by chipifying all of them.
 */
class ChipifyingNachoValidator : NachoValidator {

    override fun isValid(@NonNull chipTokenizer: ChipTokenizer, text: CharSequence): Boolean {
        // Text is valid when there are no unterminated tokens.
        val unterminated = chipTokenizer.findAllTokens(text)
        return unterminated.isEmpty()
    }

    override fun fixText(@NonNull chipTokenizer: ChipTokenizer, invalidText: CharSequence): CharSequence {
        val builder = SpannableStringBuilder(invalidText)
        chipTokenizer.terminateAllTokens(builder)
        return builder
    }
} 